internal/kafka/internal/migrations/20221129121515_add_expires_at_to_kafka_request.go

```diff
@@ -0,0 +1,24 @@
+package migrations
+
+import (
+	"time"
+
+	"github.com/go-gormigrate/gormigrate/v2"
+	"gorm.io/gorm"
+)
+
+func addExpiresAtToKafkaRequest() *gormigrate.Migration {
+	type KafkaRequest struct {
+		ExpiresAt time.Time `json:"expires_at"`
+	}
+
+	return &gormigrate.Migration{
+		ID: "20221129121515",
+		Migrate: func(tx *gorm.DB) error {
+			return tx.AutoMigrate(&KafkaRequest{})
```

Comment:
The lifespanSeconds that this value will be based on for developer instances is in the supported_instance_types config and values currently differ in stage and production.  Didn't include it as part of this migration as this will be set by one of the Kafka reconcilers anyway ([MGDSTRM-10201](https://issues.redhat.com/browse/MGDSTRM-10201)). Does this sound ok to you @miguelsorianod @ziccardi?

Reply:
It will be set by one of the reconcilers for **new** developer instances but not for the already existing ones

Reply:
I would've thought the reconciler should handle setting that field as long as its not set for developer instances :thinking:  My understanding is that even if the lifespanSeconds or grace period was changed, KFM should not update Kafka instances if the `expires_at` was already set.

I've just realized that we return the expires_at field with a value for developers in the response from Kafka create. 
![image](https://user-images.githubusercontent.com/9078522/204591774-c849d521-23a5-4c81-b8b4-ea3ee393c5a3.png)

If we want to keep the same behaviour, we might want to also set this on Kafka creation for developer instances and a reconciler would probably not be needed (can close issue [MGDSTRM-10201](https://issues.redhat.com/browse/MGDSTRM-10201))

 In this case, setting this for existing kafka instances in this migration would make sense.

Reply:
> My understanding is that even if the lifespanSeconds or grace period was changed, KFM should not update Kafka instances if the expires_at was already set

That's my understanding too

Reply:
>If we want to keep the same behaviour, we might want to also set this on Kafka creation for developer instances and a reconciler would probably not be needed (can close issue [MGDSTRM-10201](https://issues.redhat.com/browse/MGDSTRM-10201))
> In this case, setting this for existing kafka instances in this migration would make sense if possible.

It is possible that a developer kafka could be created just after the migration has rolled out and just before all the new pods (with the new logic) are rolled out. And this developer Kafka might not have its `expires_at` column set by the migration. If we do not have a reconciler logic to handle for this case, then we could be in an inconsitent state for this kafka.

Reply:
@machi1990 that's a good point. I think we may need to do this migration in multiple steps. 

1. Add the `expires_at` column to the kafka_requests table
    - Field should be set empty (should be ok since it's not used at the moment)
2. Add logic to set the `expires_at` value for all instances with a limited lifespan (currently developer and long lived instances)
    - It should only set it if expires_at is empty.
    - Done as part of [MGDSTRM-10201](https://issues.redhat.com/browse/MGDSTRM-10201)
3. Refactor logic to use this `expires_at` value. Following steps can be done simultaneously.
    - Refactor reconcile logic for deleting expired instances
    - Refactor KafkaRequest presenter so that it takes the `expires_at` value from the db instead of dynamically calculating it.
    - Refactor Kafka creation so that it sets the `expires_at` value on create if lifespanSeconds was defined for that instance type.
       - This keeps the existing behaviour of returning the `expires_at` with a value for developer instances on create. If we don't do this here, `expires_at` value will be empty for a few moments until the reconciler sets it.  
4. Remove reconcile logic to set `expires_at` value for developer instances if no longer needed.

wdyt? does this make sense? :thinking:

Reply:
The steps above should be able to handle this migration perfectly. 
Thanks for describing the proposal and welcome back :-)